### ISSUE TITLE
`-Wsometimes-uninitialized` false positive for assignment

### ISSUE THREAD

**Issue by 2e1c2d01-a631-41c2-93fe-8d40b95d8607** [llvm/llvm-project#49580]
|  |  |
| --- | --- |
| Bugzilla Link | [50236](https://llvm.org/bz50236) |
| Version | 12.0 |
| OS | Linux |
| Attachments | [Repro](https://user-images.githubusercontent.com/95090385/143762108-be127cf4-e75e-423a-b6d2-c52757d73b28.gz) |
| CC | @dwblaikie |

## Extended Description 
```
int foo(int *);

int bar(int c)
{
  int a, b;
  if (!c) c = foo(&a);
  b = a;
  if (c) return 1;
  return b;
}
```
Compiling the attached C source (named sometimes.c) gives the following warning:
```
sometimes.c:8:7: warning: variable 'a' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
  if (!c) c = foo(&a);
      ^~
sometimes.c:9:7: note: uninitialized use occurs here
  b = a;
      ^
sometimes.c:8:3: note: remove the 'if' if its condition is always true
  if (!c) c = foo(&a);
  ^~~~~~~~
sometimes.c:7:8: note: initialize the variable 'a' to silence this warning
  int a, b;
       ^
        = 0
1 warning generated.
```
At line 9, the value of sometimes-uninitialized variable `a` is only assigned to another variable `b`, and so it should not be warned until the value is actually used.

**Comment by Quuxplusone** [comment]
Yeah, `int a, b; b = a;` is absolutely undefined behavior. The extra control flow around `c` is a red herring. You can reduce this down to
```
void bar() {
  int a, b;
  b = a;  // UB here
  (void)b;
}
```
Clang is correct to warn.
